### ISSUE TITLE
Ship pathfinding issue with panama-like cities.

### ISSUE THREAD

**Issue by abmpicoli2** [We-the-People-civ4col-mod/Mod#4]
Original discussion here: https://forums.civfanatics.com/threads/religion-and-revolution-bugs-and-todos.449958/page-123#post-15217845 . FOr cities that can travel both east and west, the pathfinding from europe seems to prefer that ships go through west instead of east.

**Comment by abmpicoli** [comment]
Also discussed at this forum topic. https://forums.civfanatics.com/threads/2-7-1-bug-ships-taking-the-longer-west-trip-to-the-new-world-for-panama-like-colonies.636459/

**Comment by Nightinggale** [comment]
In short (if I get this issue correctly and I think I do):
CvPlot has a cached distance to Europe. Map setup will then init the cache and it's final. It never changes.

Proposal (which would fix this)
Make an array of cached distances in CvPlot, one for each player, or possibly one for each European player (I think they are always first in the array, but we need to verify this somehow).

If not all players have the same cache, the city owner and friends of the city owner can calculate new distances where the city is counted as a water plot. This should fix the problem.

With a player specific setup, we would also be able to mark "don't pass" for other purposes like through enemy waters during war. We can even make something like the sign code where players can set blocked plots for auto paths. Think of the controls like placing a sign. That's code which allows clicking on a plot and then we know which plot is being clicked. There is also code to draw art like circles on the plots with minimal coding meaning we can mark used selected blocked plots. The art (I think we can pick more than just circles) can be of any color, meaning we can use map markers for a lot of stuff if we want to.

**Comment by ShadesOT** [comment]
By chance I stumbled accross this very issue in my test game. = Savegame that shows it. Whoop! Use the ship headed for jamestown. Send it to europe. It will travel the west route.

[AutoSave_AD-1508.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/2503879/AutoSave_AD-1508.zip)

**Comment by Nightinggale** [comment]
Looks like the issue is in CvUnitAI::AI_sailToEurope(), more specifically
```cpp
if (generatePath(pLoopPlot, MOVE_BUST_FOG, true, &iPathTurns))
{
	int iValue = 10000;
	iValue /= 100 + getPathCost();
```
getPathCost() takes the path to the plot into consideration, but not travel time to Europe. It needs something like:
```cpp
int iTripLength = GC.getEuropeInfo(pLoopPlot->getEurope()).getTripLength();
```
The question is how to add it to make sense here. I can't really figure out how the internal move cost numbers are supposed to work here.

This would only solve part of the problem as it doesn't help with "Europe->Jamestown" pathfinding.

Also the code here is a mess. It looks like there are 3-4 functions in a row, which does the same thing.

**Comment by Nightinggale** [comment]
This would likely be a 2.8 fix rather than 2.7.x. It appears to be a not insignificant amount of work, meaning it's likely not realistic to make a good solution within the timeframe we want for 2.7.2. Also with an extra cache, it's an open question if we want to alter the savegame itself.

As for the solution proposal, I think we should make the cache team specific rather than player specific.

**Comment by raystuttgart** [comment]
This is outdated.
We already have other issues related to Pathfindings as well.
Thus I close this one.
